### ISSUE TITLE
Consider using SIMD primitives to accelerate computation

### ISSUE THREAD

**Issue by mssun** [sjtu-yogcrypt-team/yogcrypt#16]
Since the basics of SIMD has arrived in Rust 1.27 stable: https://blog.rust-lang.org/2018/06/21/Rust-1.27.html#simd, we should consider to replace our `basic::cell::U64x4` with `std::simd::u64x4`. Although `std::simd` is still in nightly, I think it will be stabilized very quickly.

Pros:
  - SIMD is fast
  - it also in the `core` library which means "no-std
  - it provides handful APIs, and we don't have to maintain these by ourselves

Cons:
  - still not stabilized, we should use nightly Rust to compile
  - not sure it's compatibility
  - we should rewrite some code

**Comment by Messjer** [comment]
Thanks for the suggestion. But I believe there are no APIs for addition (nor multiplication) with carrying. Our implementation for `U64x4` represents a 256-bit number, which does not fit into the original purpose of `std::simd::u64x4` (being only a vector of 4 `u64`'s). Therefore we cannot quite use the APIs for our big integers implementation (not in a direct and efficient way). What is your opinion on this problem?

**Comment by mssun** [comment]
Do you suggest to use basic SIMD ops (https://github.com/rust-lang/rust/pull/49664/) to speedup the code?

**Comment by Messjer** [comment]
Not really. I mean they are suitable when the additions can be run in parallel, but in our situation (big number addition) we need to consider various cases for handling carrying of the different blocks. So `std::simd::u64x4` cannot replace our implementation for `U64x4` in a direct way.

**Comment by mssun** [comment]
Yes, forget about `std::simd::u64x4`. I mean we can use basic SIMD primitives to accelerate other operations.

**Comment by Messjer** [comment]
Yes, SIMD might be used in the following use cases.
```
//ecc_group.rs L.364 - 367
let lambda3 = lambda1 - lambda2;
let lambda6 = lambda4 - lambda5;
let lambda7 = lambda1 + lambda2;
let lambda8 = lambda4 + lambda5;
```
However, if we swtich to SIMD, we may need to provide other APIs (for 4-way parallel addition) and/or breaking the `U64x4` abstraction.

**Comment by mssun** [comment]
Right, this is one potential way for optimization.
